### PR TITLE
ARROW-5458: [C++] Apache Arrow parallel CRC32c computation optimization

### PR DESCRIPTION
**Pull request by guyuqi**
ARMv8 defines VMULL/PMULL crypto instruction.
This patch optimizes crc32c calculate with the instruction when
available rather than original linear crc instructions.

### CODE REVIEW

**Comment by emkornfield** [comment]
@guyuqi thank you for the contribution.  Could you clarify if this code is an original contribution or if it was borrowed from another project? 

 If this was borrowed from someplace we need to make it has an Apache compatible license and update LICENSE.txt file to note the source.

**Comment by pitrou** [comment]
I have some questions:
- did you benchmark this on a actual workload? It seems this implementation would only be faster on values larger than 1kB, and this seems highly unlikely (except perhaps an array of large strings)
- where does the original algorithm come from? Could you add a reference somewhere?

**Comment by guyuqi** [comment]
The algorithm comes from the intel whitepaper:  [crc-iscsi-polynomial-crc32-instruction-paper](https://www.intel.com/content/dam/www/public/us/en/documents/white-papers/crc-iscsi-polynomial-crc32-instruction-paper.pdf) .
We implement it with Arm64 crypto extension.
The performance gets boost on a big size of data set.

**Comment by pitrou** [comment]
> The performance gets boost on a big size of data set.

Can you elaborate? What did you test it on?

**Comment by emkornfield** [comment]
@guyuqi this code looks very similar to the code [here](https://github.com/ceph/ceph/blob/master/src/common/crc32c_aarch64.c) did you adapt it from the ceph code?

**Comment by guyuqi** [comment]
I run hashing-benchmark:

Software crc32:
```
----------------------------------------------------------------------
BM_HashIntegers/repeats:1              20 us         20 us      34273   7.29716GB/s   934.036M items/s
BM_HashSmallStrings/repeats:1         192 us        192 us       3648    1083.6MB/s    99.187M items/s
BM_HashMediumStrings/repeats:1        717 us        717 us        979   1.80917GB/s   26.6173M items/s
BM_HashLargeStrings/repeats:1         706 us        706 us       1006   2.71318GB/s   2.70067M items/s
```

Arm64 linear crc32:
```
----------------------------------------------------------------------
BM_HashIntegers/repeats:1              23 us         23 us      29969   6.38098GB/s   816.765M items/s
BM_HashSmallStrings/repeats:1         186 us        186 us       3769   1122.15MB/s   102.716M items/s
BM_HashMediumStrings/repeats:1        551 us        550 us       1271   2.35515GB/s     34.65M items/s
BM_HashLargeStrings/repeats:1         356 us        356 us       1952   5.38805GB/s   5.36322M items/s
```


Arm64 parallel crc32 optimization:
```
----------------------------------------------------------------------
BM_HashIntegers/repeats:1              21 us         21 us      33959    7.2272GB/s   925.082M items/s
BM_HashSmallStrings/repeats:1         193 us        193 us       3632   1078.44MB/s   98.7151M items/s
BM_HashMediumStrings/repeats:1        556 us        556 us       1261   2.33304GB/s   34.3247M items/s
BM_HashLargeStrings/repeats:1         322 us        322 us       2153   5.94568GB/s   5.91828M items/s

```

 I also try to design a direct benchmark for CRC32.
Benchmark snippet( Test data size : 1M):
```
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdint.h>
#include <sys/time.h>
 
long int GetTickCount() {
        struct timeval tv;
        gettimeofday(&tv, NULL);
        return tv.tv_sec * 1000000 + tv.tv_usec;
}
 
int main() {
        static const uint64_t kSize = 1024 * 1024 + 29;
        uint8_t* buf = (uint8_t *)malloc(sizeof(uint8_t) * kSize);
        uint32_t i;
 
        srand(0);
        for (i = 0; i < kSize; i++) {
                buf[i] = (uint8_t)(rand() % 256u);
        }
 
        uint32_t kLoop = 1024;
        long int start, end;
 
        uint32_t crc = 0;
 
        start = GetTickCount();
        for (i = 0; i < kLoop; i++) {
            crc = Armv8CrcHashParallel (buf, kSize, crc);
            //crc = ArmCrc32c_linear (buf, kSize, crc);
            //crc = crc32c_sw (buf, kSize, crc);
        }
        end = GetTickCount();
 
        if (kSize < 20) {
                for (i = 0; i < kSize; i++) {
                        printf("%3u,", (uint32_t)buf[i]);
                }
                printf("\n");
        }
 
        printf("crc result = %x, time cost per loop:%f ms\n", crc, (double)(end - start) / kLoop);
 
        free(buf);
        return 0;
}
```
Data Size 1M:
```
( Cavium)
crc32c software:         1035.202ms
Arm crc32c:                232.984ms
Arm crc32c parallel:   115.589ms.
```


@emkornfield 
The code implementation is porting from ceph crc32c_aarch64.c  which was contributed by the Arm colleague(@weixiaowilliam) from our team.  We also contributed the parallel crc32c code to MariaDB([PR](https://github.com/MariaDB/server/pull/772)) last year.
Of course, I will add annotation in code to show where it comes from.

**Comment by pitrou** [comment]
So at best there's a 10% improvement in `BM_HashLargeStrings`. That's a micro-benchmark, so gains in real-world composite workloads will probably be far smaller. I'm not sure it justifies adding this. @wesm @fsaintjacques Opinions?

**Comment by emkornfield** [comment]
@guyuki those two projects are LGPL which isn't directly compatible with the Apache license.  I think we would need the original copyright holder to give explicit permission to relicense it (@pitrou can you confirm?)  Can you clarify if you or your colleague are the original authors?

**Comment by guyuqi** [comment]
In the patch: 
if data size larger than 1024, the code will go to linear crc32c computation.
```
...
..
while ((nbytes -= 1024) >= 0) {
...
..
.
}
```
From my benchmark , we will get ~ x2.01 speedups.

**Comment by wesm** [comment]
@emkornfield correct -- reusing LGPL code is not permissible in this fashion unless the work is being contributed under Apache 2.0 by the copyright holders

**Comment by wesm** [comment]
@guyuqi we need clarification about licensing and IP origin before we can consider merging this

**Review by wesm — CHANGES_REQUESTED** [review]
@guyuqi we need clarification about licensing and IP origin before we can consider merging this

**Comment by guyuqi** [comment]
@wesm 
The code is re-implemented with the algorithm derived from Intel whitepaper.
The useless code of L2/L1 cache prefetching was also removed.

**Comment by pitrou** [comment]
Can you post results for hashing-benchmark again?

File: cpp/src/arrow/util/hash-util.h [hunk 1]
```diff
@@ -72,80 +72,80 @@ class HashUtil {
...
-      /* Do first 8 bytes here for better pipelining */
-      crc0 = __crc32cd(crc, *(const uint64_t*)buffer);
-      crc1 = 0;
-      crc2 = 0;
-      buffer += sizeof(uint64_t);
-
-      /* Process block inline
-       * Process crc0 last to avoid dependency with above
+    const uint8_t* buf8;
+    const uint64_t* buf64 = (const uint64_t*)data;
```
Comment (wesm) [inline cpp/src/arrow/util/hash-util.h]: Can you use C++-style casts in this file (`reinterpret_cast` and `static_cast`)?

**Comment by guyuqi** [comment]
Arrow-hashing-benchmark:

```
---------------------------------------------------------
Benchmark                  Time           CPU Iterations
---------------------------------------------------------
HashIntegers           21140 ns      21137 ns      33196   7.04989GB/s   902.385M items/s
HashSmallStrings      195629 ns     195605 ns       3627   1065.28MB/s   97.5103M items/s
HashMediumStrings     551390 ns     551311 ns       1270   2.35152GB/s   34.5966M items/s
HashLargeStrings      322247 ns     322049 ns       2152   5.94996GB/s   5.92253M items/s
```

File: cpp/src/arrow/util/hash-util.h (hunk 1 above)
Comment (guyuqi) [reply cpp/src/arrow/util/hash-util.h]: OK. Done.

**Comment by pitrou** [comment]
So this is winning 10% on a micro-benchmark that's not representative of any real-world workload. I'm skeptical about maintaining this in our codebase for such a little gain.

**Comment by pitrou** [comment]
For a more general perspective, we don't actually care about computing a CRC here. We're just using CRC instructions to devise a good, fast, hash function. See `DoubleCrcHash`: it's not an actual CRC, just a hack to exploit CRC computation parallelism on the CPU.

**Comment by wesm** [comment]
A flip side of the argument is that having this algorithm provides a sound Apache2-compatible implementation which can hopefully be maintained by other ARM users (or we can remove it later if it grows burdensome)

File: cpp/src/arrow/util/hash-util.h [hunk 2]
```diff
@@ -71,6 +71,86 @@ class HashUtil {
...
+#ifdef ARROW_HAVE_ARMV8_CRYPTO
+/* Crc32c Parallel computation
+ *   Algorithm comes from Intel whitepaper:
+ *   crc-iscsi-polynomial-crc32-instruction-paper
+ *
+ * Input data is divided into three equal-sized blocks
+ *   Three parallel blocks (crc0, crc1, crc2) for 1024 Bytes
+ *   One Block: 42(BLK_LENGTH) * 8(step length: crc32c_u64) bytes
+ */
+#define BLK_LENGTH 42
```
Comment (emkornfield) [inline cpp/src/arrow/util/hash-util.h]: use constexpr

Comment (emkornfield) [reply cpp/src/arrow/util/hash-util.h]: kBlockLength

**Comment by fsaintjacques** [comment]
@wesm are you ok with the IP answer?

**Review by wesm — APPROVED** [review]
+1

**Pull request closed without merging**